From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DiaLight <light_01@rambler.ru>
Date: Thu, 11 Jul 2024 03:10:55 +0300
Subject: [PATCH] use_cwd_as_dk2_home_dir


diff --git a/dk2/MyResources.cpp b/dk2/MyResources.cpp
index 0000000..0000000 100644
--- a/dk2/MyResources.cpp
+++ b/dk2/MyResources.cpp
@@ -44,6 +44,11 @@ dk2::MyResources *dk2::MyResources::init_resources() {
         if (str_end > cmdl) str_end[1] = '\0';
     }
     _strcpy(exeDir, cmdl);
+    // patch use_cwd_as_dk2_home_dir
+    GetCurrentDirectoryA(MAX_PATH, exeDir);
+    strcat(exeDir, "\\");
+    printf("replace exe dir path2: %s -> %s\n", cmdl, exeDir);
+    // end of patch use_cwd_as_dk2_home_dir
     LABEL_12:
     MyGame_debugMsg(&MyGame_instance, "HD Path: %s\n", exeDir);
     _strcpy(this->executableDir, exeDir);
diff --git a/dkii_exe_functions.cpp b/dkii_exe_functions.cpp
index 0000000..0000000 100644
--- a/dkii_exe_functions.cpp
+++ b/dkii_exe_functions.cpp
@@ -16,6 +16,14 @@ int32_t dk2::MyGame::isOsCompatible() {
 }
 
 void dk2::resolveDk2HomeDir() {
+    // patch use_cwd_as_dk2_home_dir
+    char tmp[MAX_PATH];
+    DWORD len = GetCurrentDirectoryA(MAX_PATH, tmp);
+    strcpy(tmp + len, "\\");
+    printf("replace exe dir path1: %s -> %s\n", dk2::dk2HomeDir, tmp);
+    strcpy(dk2::dk2HomeDir, tmp);
+    return;
+    // end of patch use_cwd_as_dk2_home_dir
     const char *CommandLineA = GetCommandLineA();
     _strncpy(pathBuf, CommandLineA, 259u);
     char firstChar = pathBuf[0];
