From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DiaLight <light_01@rambler.ru>
Date: Thu, 11 Jul 2024 03:20:43 +0300
Subject: [PATCH] use_wheel_to_zoom


diff --git a/dkii_exe_functions.cpp b/dkii_exe_functions.cpp
index 0000000..0000000 100644
--- a/dkii_exe_functions.cpp
+++ b/dkii_exe_functions.cpp
@@ -93,6 +93,34 @@ LRESULT dk2::CWindowTest_proc(HWND hWnd, UINT Msg, WPARAM wParam, LPARAM lParam)
         }
     }
     // end of patch fix_mouse_pos_on_resized_window
+    // patch use_wheel_to_zoom
+    switch (Msg) {
+        case WM_MOUSEWHEEL: {
+            DWORD fwKeys = GET_KEYSTATE_WPARAM(wParam);
+            DWORD zDelta = GET_WHEEL_DELTA_WPARAM(wParam);  // +-120*speed
+            DWORD xPos = GET_X_LPARAM(lParam);
+            DWORD yPos = GET_Y_LPARAM(lParam);
+//            printf("k=%08X d=%d {%d %d}\n", fwKeys, zDelta, xPos, yPos);
+            dk2::CBridge_instance.camera.zoomRel_449CA0(-zDelta * 50);
+            break;
+        }
+    }
+//    + hook::DIRECT_INPUT_MOUSE_DATA.emplace_back([](DIDEVICEOBJECTDATA *data) {
+//        switch (data->dwOfs) {
+//            case DIMOFS_Z: {  // mouse wheel
+//                int zDelta = data->dwData;  // +-150 with timestamp
+//                int tsDelta = data->dwTimeStamp - g_lastTimestamp;
+//                g_lastTimestamp = data->dwTimeStamp;
+////                printf("wheel: %d, ts: %d\n", data->dwData, tsDelta);
+//                int mult = 80;
+//                if (tsDelta > 100) mult = 40;
+//                if (tsDelta > 500) mult = 20;
+//                zoom2(-zDelta * mult);
+//                break;
+//            }
+//        }
+//    });
+    // end of patch use_wheel_to_zoom
     // patch fix_keyboard_state_on_alt_tab
     switch(Msg) {
         case WM_ACTIVATEAPP:
