From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DiaLight <light_01@rambler.ru>
Date: Thu, 11 Jul 2024 03:21:45 +0300
Subject: [PATCH] replace_mouse_dinput_to_user32


diff --git a/dk2/MyDxMouse.cpp b/dk2/MyDxMouse.cpp
index 0000000..0000000 100644
--- a/dk2/MyDxMouse.cpp
+++ b/dk2/MyDxMouse.cpp
@@ -14,6 +14,10 @@ int *dk2::MyDxInputManagerCb::initMouse(int *pstatus) {
     }
     int status;
     MyDxMouse_create(&status, &this->f58_pdxmouse);
+    // patch replace_mouse_dinput_to_user32
+    *pstatus = STATUS_SUCCESS;
+    return pstatus;
+    // end of patch replace_mouse_dinput_to_user32
     if (FAILED(status)) {
         *pstatus = status;
         return pstatus;
@@ -30,6 +34,11 @@ int *dk2::MyDxInputManagerCb::initMouse(int *pstatus) {
 }
 
 int *__cdecl dk2::MyDxMouse_create(int *pstatus, MyDxMouse **pObj) {
+    // patch replace_mouse_dinput_to_user32
+    *pObj = nullptr;
+    *pstatus = STATUS_SUCCESS;
+    return pstatus;
+    // end of patch replace_mouse_dinput_to_user32
     if (pObj == nullptr) {
         *pstatus = 0x80004003;
         return pstatus;
