From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DiaLight <light_01@rambler.ru>
Date: Thu, 11 Jul 2024 03:21:56 +0300
Subject: [PATCH] control_windowed_mode


diff --git a/dk2/MyGame.cpp b/dk2/MyGame.cpp
index 0000000..0000000 100644
--- a/dk2/MyGame.cpp
+++ b/dk2/MyGame.cpp
@@ -16,6 +16,10 @@ int dk2::MyGame::prepareScreenEx(
         int isWindowed,
         int screenSwap,
         int screenHardware3D) {
+    // patch control_windowed_mode
+    printf("prepareScreen %p %dx%d %d %d %d %d\n", this, dwWidth, dwHeight, dwRGBBitCount, isWindowed, screenSwap, screenHardware3D);
+    isWindowed = true;  // todo: control
+    // end of patch control_windowed_mode
     int sel_dd_idx = this->selected_dd_idx;
     if (sel_dd_idx != this->last_selected_dd_idx) {
         MyResources_instance.video_settings.sub_566E40(sel_dd_idx);
